[Quant] Enable default reference path for CopyNodeQuantizeHandler

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71168

Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33715995](https://our.internmc.facebook.com/intern/diff/D33715995)

terrychenism committed:
[Quant] Enable default reference path for CopyNodeQuantizeHandler

Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33534482).

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33534482).

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33534482).

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1555,9 +1555,17 @@ def convert(self,
+            'relu',
+            'relu_',
+        ]
+        is_func = node.op == "call_function" and node.target in func_list
```
is_call_method = node.op == "call_method" and node.target in method_list
is_call_module = node.op == "call_module" and type(modules[node.target]) in module_type_list

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -78,6 +78,85 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule, ref_class: Type[torc
+                    zero_point_node = q_node.args[2]
+                    dtype = q_node.args[3]
+
+                    output_scale = getattr(model, scale_node.target)
+                    output_zero_point = getattr(model, zero_point_node.target)
```
these are not used I think

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33534482](https://our.internmc.facebook.com/intern/diff/D33534482)

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -78,6 +78,82 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule, ref_class: Type[torc
+def special_pattern_replacement(model: QuantizedGraphModule) -> QuantizedGraphModule:
+    modules = dict(model.named_modules(remove_duplicate=False))
+    nodes = list(model.graph.nodes)
+    module_type_list = [
```
Also maybe turn these into sets to avoid linear lookup per node?

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -78,6 +78,82 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule, ref_class: Type[torc
+                    # get output scale/zero_point/dtype from the quantize node
+                    scale_node = q_node.args[1]
+                    zero_point_node = q_node.args[2]
+                    dtype = q_node.args[3]
```
nit: I think this is more concise
```
ref_node, scale_node, zero_point_node, dtype = q_node.args
```

andrewor14 commented on torch/ao/quantization/fx/_lower_to_native_backend.py:
```diff
@@ -78,6 +78,82 @@ def _lower_weighted_ref_module(model: QuantizedGraphModule, ref_class: Type[torc
+                    else:
+                        dq_node.target = ref_node
+
+                    # remvoe dq node:
```
This part is copy and pasted from `_lower_weighted_ref_module`. Can you make a helper function? Also typo: remvoe

terrychenism replied:
this part is slightly different from the function in _lower_weighted_ref_module

andrewor14 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -1555,9 +1555,41 @@ def convert(self,
+        ]
+        is_call_function = node.op == "call_function" and node.target in func_list
+        is_call_method = node.op == "call_method" and node.target in method_list
+        is_call_module = node.op == "call_module" and type(modules[str(node.target)]) in module_type_list
```
This part is duplicated too

andrewor14 reviewed (commented):
Hi @terrychenism, this looks great! I left some minor comments mostly related to reducing duplicate code.

andrewor14 commented:
Also should we add lowering tests for each of the ops or are the existing tests fine? @jerryzh168

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:



[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33715995).

jerryzh168 commented:
> Also should we add lowering tests for each of the ops or are the existing tests fine? @jerryzh168

existing tests should be OK I think, these are tested in https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L4826 and https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L4733, but we could split them later to smaller tests in the future

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33715995](https://our.internmc.facebook.com/intern/diff/D33715995)

[ghstack-poisoned]

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33715995](https://our.internmc.facebook.com/intern/diff/D33715995)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33715995).

terrychenism committed:
Update on "[Quant] Enable default reference path for CopyNodeQuantizeHandler"


Summary:
In this PR we want to enable the reference path by default for CopyNodeQuantizeHandler

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33715995](https://our.internmc.facebook.com/intern/diff/D33715995)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33715995).

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -12,6 +12,42 @@
+    'relu_',
+}
+
+def check_node(node, modules):
```
can we move this function to a less general place.. like _lower_to_native_bakcend.py
